src/api_manager/service_control/aggregated.cc

```diff
@@ -548,6 +520,47 @@ const std::string& Aggregated::GetAuthToken<AllocateQuotaRequest>() {
   }
 }
 
+template <>
+void Aggregated::HandleResponse(const CheckResponse& response) {
+  if (set_rollout_id_func_ && !response.service_rollout_id().empty()) {
+    set_rollout_id_func_(response.service_rollout_id());
+  }
+
+  if (!response.service_config_id().empty() &&
+      service_control_proto_.service_config_id() !=
+          response.service_config_id()) {
+    if (mismatched_check_config_id_ != response.service_config_id()) {
+      env_->LogDebug(
+          "Received non-matching check response service config ID: '" +
+          response.service_config_id() + "', requested: '" +
+          service_control_proto_.service_config_id() + "'");
+      mismatched_check_config_id_ = response.service_config_id();
+    }
+  }
+}
+
+template <>
+void Aggregated::HandleResponse(const ReportResponse& response) {
```

Comment:
Duplicated?

Reply:
Not, it is a template function with different type.

Reply:
Hmm so it seems like the body is identical for CheckResponse and ReportResponse.
Can we only specialize AllocateQuotaResponse?

i.e. use regular template for CheckResponse and ReportResponse, and use template specialization for AllocateQuotaResponse.

WDYT?

Reply:
No,  there is subtle different.  Each is using a different member variable mismatched_check_config_id_  for Check and mismatched_report_config_id_ for Report

If we want to get fancier,  we could use a template member variable. But it is only supported in c++14,  ESP is using c++11.

Or we could combine them into one mismatched_config_id_ for Check and Report.  But It may toggle, the idea is not reduce the warning log if config_id used in esp is different than the one used in the server.   Check and Report may go to different servers, so they may be using different config_id

Reply:
I see. SG